File: Extensions/SciPy/cminpack-master/examples/ibmdpdr_.c
#include <stdio.h>
#include <math.h>
#include "minpack.h"
#define real __minpack_real__

extern void machar_(int *ibeta, int *it, int *irnd, 
                    int *ngrd, int *machep, int *negep, int *iexp, 
                    int *minexp, int *maxexp, real *eps, real *epsneg,
                    real *xmin, real *xmax);

/*     ********** */

/*     this program checks the constants of machine precision and */
/*     smallest and largest machine representable numbers specified in */
/*     function dpmpar, against the corresponding hardware-determined */
/*     machine constants obtained by machar, a subroutine due to */
/*     w. j. cody. */

/*     data statements in dpmpar corresponding to the machine used must */
/*     be activated by removing c in column 1. */

/*     the printed output consists of the machine constants obtained by */
/*     machar and comparisons of the dpmpar constants with their */
/*     machar counterparts. descriptions of the machine constants are */
/*     given in the prologue comments of machar. */

/*     subprograms called */

/*       minpack-supplied ... machar,dpmpar */

/*     argonne national laboratory. minpack project. march 1980. */
/*     burton s. garbow, kenneth e. hillstrom, jorge j. more */

/*     ********** */
/* Main program */
int main(int argc, char **argv)
{
    /* Local variables */
    int it;
    real eps;
    int ngrd, irnd, iexp;
    real rerr[3], xmin, xmax;
    int ibeta, negep;
    real giant, dwarf;
    int machep;
    real epsmch, epsneg;
    int minexp, maxexp;
    const int i1 = 1, i2 = 2, i3 = 3;
    (void)argc; (void)argv;

/*     determine the machine constants dynamically from machar. */

    machar_(&ibeta, &it, &irnd, &ngrd, &machep, &negep, &iexp, &minexp, &
	    maxexp, &eps, &epsneg, &xmin, &xmax);

/*     compare the dpmpar constants with their machar counterparts and */
/*     store the relative differences in rerr. */

    epsmch = __minpack_func__(dpmpar)(&i1);
    dwarf = __minpack_func__(dpmpar)(&i2);
    giant = __minpack_func__(dpmpar)(&i3);
    rerr[0] = (epsmch - eps) / epsmch;
    rerr[1] = (dwarf - xmin) / dwarf;
    rerr[2] = (xmax - giant) / giant;

/*     write the MACHAR constants. */

    printf("\f MACHAR constants\n\n\n");
    printf(" ibeta =%6i\n\n", ibeta);
    printf(" it    =%6i\n\n", it);
    printf(" irnd  =%6i\n\n", irnd);
    printf(" ngrd  =%6i\n\n", ngrd);
    printf(" machep =%6i\n\n", machep);
    printf(" negep =%6i\n\n", negep);
    printf(" iexp =%6i\n\n", iexp);
    printf(" minexp =%6i\n\n", minexp);
    printf(" maxexp =%6i\n\n", maxexp);
    printf(" eps =%15.7e\n\n", (double)eps);
    printf(" epsneg =%15.7e\n\n", (double)epsneg);
    printf(" xmin =%15.7e\n\n", (double)xmin);
    printf(" xmax =%15.7e\n\n", (double)xmax);


/*     write the DPMPAR constants and the relative differences. */

    printf("\n\n DPMPAR constants and relative differences\n\n\n");
    printf(" epsmch =%15.7e\n", (double)epsmch);
    printf(" rerr(1) =%15.7e\n\n", (double)rerr[0]);
    printf(" dwarf =%15.7e\n", (double)dwarf);
    printf(" rerr(2) =%15.7e\n\n", (double)rerr[1]);
    printf(" giant =%15.7e\n", (double)giant);
    printf(" rerr(3) =%15.7e\n", (double)rerr[2]);

    return 0;
} /* MAIN__ */

File: Extensions/SciPy/cminpack-master/examples/ibmdpdr.f
c     **********
c
c     this program checks the constants of machine precision and
c     smallest and largest machine representable numbers specified in
c     function dpmpar, against the corresponding hardware-determined
c     machine constants obtained by machar, a subroutine due to
c     w. j. cody.
c
c     data statements in dpmpar corresponding to the machine used must
c     be activated by removing c in column 1.
c
c     the printed output consists of the machine constants obtained by
c     machar and comparisons of the dpmpar constants with their
c     machar counterparts. descriptions of the machine constants are
c     given in the prologue comments of machar.
c
c     subprograms called
c
c       minpack-supplied ... machar,dpmpar
c
c     argonne national laboratory. minpack project. march 1980.
c     burton s. garbow, kenneth e. hillstrom, jorge j. more
c
c     **********
      integer ibeta,iexp,irnd,it,machep,maxexp,minexp,negep,ngrd,
     *        nwrite
      double precision dwarf,eps,epsmch,epsneg,giant,xmax,xmin
      double precision rerr(3)
      double precision dpmpar
c
c     logical output unit is assumed to be number 6.
c
      data nwrite /6/
c
c     determine the machine constants dynamically from machar.
c
      call machar(ibeta,it,irnd,ngrd,machep,negep,iexp,minexp,maxexp,
     *            eps,epsneg,xmin,xmax)
c
c     compare the dpmpar constants with their machar counterparts and
c     store the relative differences in rerr.
c
      epsmch = dpmpar(1)
      dwarf = dpmpar(2)
      giant = dpmpar(3)
      rerr(1) = (epsmch - eps)/epsmch
      rerr(2) = (dwarf - xmin)/dwarf
      rerr(3) = (xmax - giant)/giant
c
c     write the machar constants.
c
      write (nwrite,10)
     *      ibeta,it,irnd,ngrd,machep,negep,iexp,minexp,maxexp,eps,
     *      epsneg,xmin,xmax
c
c     write the dpmpar constants and the relative differences.
c
      write (nwrite,20) epsmch,rerr(1),dwarf,rerr(2),giant,rerr(3)
      stop
   10 format (17h1MACHAR constants /// 8h ibeta =, i6 // 8h it    =,
     *        i6 // 8h irnd  =, i6 // 8h ngrd  =, i6 // 9h machep =,
     *        i6 // 8h negep =, i6 // 7h iexp =, i6 // 9h minexp =,
     *        i6 // 9h maxexp =, i6 // 6h eps =, d15.7 // 9h epsneg =,
     *        d15.7 // 7h xmin =, d15.7 // 7h xmax =, d15.7)
   20 format ( /// 42h dpmpar constants and relative differences ///
     *         9h epsmch =, d15.7 / 10h rerr(1) =, d15.7 //
     *         8h dwarf =, d15.7 / 10h rerr(2) =, d15.7 // 8h giant =,
     *         d15.7 / 10h rerr(3) =, d15.7)
c
c     last card of driver.
c
      end
